src/core/DY.Core.Trace.Base.fst

```diff
@@ -855,14 +814,79 @@ val trace_find_first: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label
 let trace_find_first tr e =
   Some?.v (trace_search_first tr (trace_entry_equiv e))
 
-[@@"opaque_to_smt"]
-val trace_find_last: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label_t{entry_exists tr e} ->
+val trace_find_last: #label_t:Type -> tr:trace_ label_t -> e:trace_entry_ label_t ->
     Pure timestamp
-      (requires True)
+      (requires equiv_entry_exists tr e)
       (ensures fun ts ->
         ts `on_trace` tr /\
         trace_entry_equiv e (get_entry_at tr ts) /\
         forall ts'. (ts' `on_trace` tr /\ trace_entry_equiv e (get_entry_at tr ts')) ==> ts' <= ts
       )
 let trace_find_last tr e =
   Some?.v (trace_search_last tr (trace_entry_equiv e))
+
+val opt_less_defined: #a:eqtype -> option a -> option a -> bool
+let opt_less_defined o1 o2 =
+  o1 = None || o1 = o2
+
+val trace_search_first_later:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  p:(trace_entry_ label_t -> bool) ->
+  Lemma
+  (requires tr1 <$ tr2)
+  (ensures opt_less_defined (trace_search_first tr1 p) (trace_search_first tr2 p))
+  [SMTPat (tr1 <$ tr2);
+   SMTPat (trace_search_first tr1 p);
+   SMTPat (trace_search_first tr2 p);
+  ]
+let trace_search_first_later tr1 tr2 p = ()
```

Comment:
Same here, interesting to see that no definition revealing is required because the theorem can de deduced from `trace_search_first` postcondition + other `_later` lemmas